mlir/lib/Dialect/SparseTensor/Pipelines/SparseTensorPipelines.cpp

```diff
@@ -78,11 +78,13 @@ void mlir::sparse_tensor::buildSparseCompiler(
 
   // Finalize GPU code generation.
   if (gpuCodegen) {
-#if MLIR_GPU_TO_CUBIN_PASS_ENABLE
-    pm.addNestedPass<gpu::GPUModuleOp>(createGpuSerializeToCubinPass(
-        options.gpuTriple, options.gpuChip, options.gpuFeatures));
-#endif
+    GpuNVVMAttachTargetOptions nvvmTargetOptions;
+    nvvmTargetOptions.triple = options.gpuTriple;
+    nvvmTargetOptions.chip = options.gpuChip;
+    nvvmTargetOptions.features = options.gpuFeatures;
+    pm.addPass(createGpuNVVMAttachTarget(nvvmTargetOptions));
     pm.addPass(createGpuToLLVMConversionPass());
+    pm.addPass(createGpuModuleToBinaryPass());
```

Comment:
@aartbik here's the change that keeps the Sparse pipeline working.

With respect to the tests, thank you! To be honest, I don't know if the problem is my test bed, it behaved erratically today.

Reply:
Thank you for the extra information (with the recent change from Phabricator to PR, I was a bit behind on all the changes). Again, your cleanup is much appreciated and I will look into this Monday. Worst case, we will fix it "after the fact", I don't want to hold up your process unnecessarily and this codepath is not very critical for us (yet ;-)

Reply:
I looked into the test, and the 

'cuMemAlloc(&ptr, sizeBytes)' failed with 'CUDA_ERROR_INVALID_VALUE'

messages are indeed prompted, but seem harmless, since they just indicate that the method is called with sizeBytes == 0. I have several places in the GPU libgen path of the sparse compiler that bumps a zero size to a size of 4 just to avoid the message.

But other than that, the GPU path continues calling cuSPARSE and producing the correct result.

( ( 1, 39, 52, 0, 0, 0, 45, 51 ), ( 0, 0, 0, 0, 0, 0, 0, 0 ), ( 0, 0, 16, 0, 0, 0, 0, 0 ), ( 0, 0, 0, 25, 0, 0, 0, 0 ), ( 0, 0, 0, 0, 36, 0, 0, 0 ), ( 0, 117, 158, 0, 0, 0, 135, 144 ), ( 0, 156, 318, 0, 0, 0, 301, 324 ), ( 0, 208, 430, 0, 0, 0, 405, 436 ) )
20

So, is the issue with cuMemAlloc(&ptr, sizeBytes) for zero size known? And should we just do something about that in the library wrapper?

@grypp WDYT?

Reply:
> So, is the issue with cuMemAlloc(&ptr, sizeBytes) for zero size known? 

Yes this is a known failure. 

> And should we just do something about that in the library wrapper?

Let's put an if statement that checks zero size in the runtime.

Reply:
Also, now that the change is in, I actually see true errors (i.e. the computed result is no longer valid), so the cuMemAlloc error message was probably a red herring here. I will look into what went wrong.

Reply:
In the same test? I'll investigate.

Reply:
I think these are currently failing on main branch in OSS:

  MLIR :: Integration/Dialect/SparseTensor/GPU/CUDA/sparse-matvec-const.mlir
  MLIR :: Integration/Dialect/SparseTensor/GPU/CUDA/sparse-matvec.mlir

with

'cuModuleLoadData(&module, data)' failed with 'CUDA_ERROR_NO_BINARY_FOR_GPU'

Reply:
(btw, https://github.com/llvm/llvm-project/pull/66127 fixes the zero size error, so the red herring should be gone now)

Reply:
By any chance, are you using the nvptxcompiler lib path? or the tool one? Because that error, means no image was generated for that GOU, which would be considered a normal error with `nvptxcompiler`.

I'm working on a fully JIT path that should solve all testing issues. I think I'll have the pr ready tomorrow.

Reply:
I have a libnvptxcompiler_static.a in my lib path when running these tests, so perhaps that is not the right setup? Still catching up on all the changes ;-), but could we generate a friendlier message for this, or is that too deeply inside the tooling to make that easy?

Reply:
No, for the nvptxcompiler path you need to pass a special CMake flag, so probably you're not using that path. So the issue must be a different one, could you send the error, the GPU model and the toolkit version?

Reply:
Very nice! Thank you!

With a proper NVPTX built MLIR, I can debug this much better with these flags. Note that the sparse compiler has a CUDA codegen path (CUDA threads are generated for outermost loops of sparse code) and a CUDA libgen path (cuSPARSE, cuSPARSElt). For the former, my desktop GPU (Quadro P1000) used to be able to run the "codegen" path with the cubin pass under the given flags, even though sm_80 was a bit too high). By adjusting the test for my desktop, I see clean run again:

Tool invocation for module: "sparse_kernels"
ptxas -arch sm_50 /tmp/mlir-sparse_kernels-nvptx64-nvidia-cuda-sm_50-05ebde.ptx -o /tmp/mlir-sparse_kernels-nvptx64-nvidia-cuda-sm_50-05ebde.ptx.cubin --opt-level 2
fatbinary -64 --image3=kind=elf,sm=50,file=/tmp/mlir-sparse_kernels-nvptx64-nvidia-cuda-sm_50-05ebde.ptx.cubin --image3=kind=ptx,sm=50,file=/tmp/mlir-sparse_kernels-nvptx64-nvidia-cuda-sm_50-05ebde.ptx --create /tmp/mlir-sparse_kernels-nvptx64-nvidia-cuda-sm_50-290575.bin

-- JIT main !

( 87360, 89440, 91520, 93600, 95680, 97760, 99840, 101920, 104000, 106080, 108160, 110240, 112320, 114400, 116480, 118560, 120640, 122720, 124800, 126880, 128960, 131040, 133120, 135200, 137280, 139360, 141440, 143520, 145600, 147680, 149760, 151840, 153920, 156000, 158080, 160160, 162240, 164320, 166400, 168480, 170560, 172640, 174720, 176800, 178880, 180960, 183040, 185120, 187200, 189280, 191360, 193440, 195520, 197600, 199680, 201760, 203840, 205920, 208000, 210080, 212160, 214240, 216320, 218400 )

Note that we still have an issue in our blaze based set up getting the  nvptxcompiler in place, but that is another story ;-)

Reply:
Oh, I see, I think I know what's happening: the tests use something like `chip=sm_80` and are running on a device with a lower compute capability (`sm_70`). 

Background: The compute capability was never utilized at any point in `--gpu-to-cubin`, and it always compiled the code for the arch found at compile time, that's why the tests never gave issues before.

With this new method, there are 2 mechanisms:
- `format=bin` compiles only for the specified chip and if there's an arch miss-match then a `CUDA_ERROR_NO_BINARY_FOR_GPU` is thrown.
-  `format=fatbin` generates a binary for the specified chip and also embeds the PTX, allowing the driver to JIT the code. However, something I found recently is that the driver can only JIT to a higher CC, so for example `chpi=sm_50` can run on `sm_80`, but `chip=sm_80` cannot run on `sm_50` and in this case one also gets `CUDA_ERROR_NO_BINARY_FOR_GPU`.

NOTE: `fatbin` is the default format.

The issue is this line [sparse-matmul-lib.mlir#L5](https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/SparseTensor/GPU/CUDA/sparse-matmul-lib.mlir#L5).

How to solve? Use the lowest arch required for the test, that's why most integration tests in trunk use `sm_50`.

Please let me know, if the above works. I think it will, because when I ran the tests on an A100  those tests passed.

Wrt to blaze, I'm working on the JIT only path, so stay tuned.

Reply:
Yeah, I changed exactly that line to

--sparse-compiler="enable-runtime-library=false parallelization-strategy=dense-outer-loop gpu-triple=nvptx64-nvidia-cuda gpu-chip=sm_50 gpu-features=+ptx60"

to make the test pass. The "problem" is that I lazily put *all* the tests under the mlir_run_cuda_sm80_tests flag, since A100s are sort of the "minimum" GPUs we are interested in. But that gave the pass-before/fail-after behavior for me (which was not really a fail since I knew my desktop was not up-to-standard). So all is good!

Reply:
Great, so thankfully the error was not a bug.